src/htbl.ml

```diff
@@ -0,0 +1,354 @@
+    let max_number_of_buckets = Int.shift_left 1 exp_max in
+
+    let first_seg =
+      Array.init segment_size (fun i ->
+          if i = 0 then Atomic.make (new_dummy_node 0 llist)
+          else Atomic.make Llist.Last)
+    in
+
+    let segments =
+      Array.init (Int.shift_left 1 grow_exp) (fun i ->
+          match i with
+          | 0 -> Atomic.make (Some first_seg)
+          | _ -> Atomic.make None)
+    in
+    {
+      count = Atomic.make 0;
+      number_of_buckets = Atomic.make segment_size;
+      segment_size;
+      max_number_of_buckets;
+      segments;
+      max;
+    }
+
+  let is_empty { count; _ } = Atomic.get count = 0
+
+  let get_segment t segment_ind =
+    let seg = t.segments.(segment_ind) in
+    match Atomic.get seg with
+    | Some seg -> seg
+    | None ->
+        (* Can failed if another domain has already done it *)
+        Atomic.compare_and_set seg None
+          (Some (Array.init t.segment_size (fun _ -> Atomic.make Llist.Last)))
+        |> ignore;
+        (* always true as either this domain or another just changed
+           the value of this cell and the only values written in
+           [t.segments] are `Some` values *)
+        Option.get (Atomic.get seg)
+
+  let get_bucket_ref t ind =
+    let segment_ind = ind / t.segment_size in
+    let segment = get_segment t segment_ind in
+    segment.(ind land (t.segment_size - 1))
+
+  (** [init_bucket buckets ind] inits a bucket and all its parents if needed *)
+  let rec init_bucket t bucket bucket_ind =
+    let parent_ind = unset_msb bucket_ind in
+    let parent_bucket = get_bucket_ref t parent_ind in
+    (if parent_ind <> bucket_ind then
+     match Atomic.get parent_bucket with
+     | Llist.Last -> init_bucket t parent_bucket parent_ind
+     | LRemove -> failwith "Should never happen"
+     | _ -> ());
+    new_dummy_node bucket_ind parent_bucket |> Atomic.set bucket
+
+  let get_bucket key t =
+    let mask_val = Atomic.get t.number_of_buckets - 1 in
+    let bucket_ind = key land mask_val in
+    let bucket = get_bucket_ref t bucket_ind in
+    match Atomic.get bucket with
+    | Llist.Last ->
+        init_bucket t bucket bucket_ind;
+        bucket
+    | LRemove -> failwith "Should never happen"
+    | _ -> bucket
+
+  let rec grow ({ number_of_buckets; max_number_of_buckets; max; _ } as t) count
+      =
+    let number_of_buckets_val = Atomic.get number_of_buckets in
+    if
+      count > number_of_buckets_val * max
+      && number_of_buckets_val < max_number_of_buckets
+    then
+      if
+        Atomic.compare_and_set t.number_of_buckets number_of_buckets_val
+          (number_of_buckets_val lsl 1)
+      then ()
+      else grow t count
+
+  (** [add key value t] *)
+  let add (key : key) (value : 'a) (t : 'a t) =
+    let is_added =
+      Llist.add (compute_hkey key) (Regular value) @@ get_bucket key t
+    in
+    if not is_added then false
+    else (
+      Atomic.fetch_and_add t.count 1 |> grow t;
```

Comment:
For the record, I did some experimenting with this FAD. It looks like a potential bottleneck since all threads doing add/remove, which are mostly disjoint accesses, get serialized here. But it does not look to matter on 4 domain benchmarks (yet). It might scale differently on arm, where fad gets turned into load/store loop.